### ISSUE TITLE
pqxx::connect not connecting

### ISSUE THREAD

**Issue by nestorcamposgestal** [jtv/libpqxx#588]
Hi, I have created my PostgreSQL database with docker compose and I am trying to connect to the database and it gives me the following error:

```
Connection refused
        Is the server running on host "127.0.0.1" and accepting
        TCP/IP connections on port 5432?
```

Here is my code:

```c++
#include <pqxx/pqxx>
using namespace pqxx;
connection C("dbname=stars_gui user=postgres password=xxx hostaddr=127.0.0.1 port=5432");
```

I tried to connect from a NodeJS application using another framework and it works fine, but from my C++ executable it does not. I've also checked that the server is listening on that port in my computer:

```
docker-pr 56101            root    4u  IPv4 278391      0t0  TCP *:5432 (LISTEN)
docker-pr 56109            root    4u  IPv6 275998      0t0  TCP *:5432 (LISTEN)
```

**Comment by jtv** [comment]
Do you get an actual error message?  Try catching any exception the connection attempt may throw.

**Comment by jtv** [comment]
Sorry, read that in too great a hurry.  It's definitely saying it's not getting a localhost connection...  That does sound as if there is _some_ difference between the network path from your NodeJS application to the database, and the network path from your C++ executable to the database.  I think you'll have to track that down somehow.

If it helps, try running `psql` instead of your C++ executable, and see if that gives  you the same problem.

**Comment by KayEss** [comment]
This depends a lot on how you're running the containers. Assuming you're running your C++ code on the host you'll need to expose your postgres container's ports to the host machine. If you're running everything in docker then it would depend on what networking options you're using in docker compose. It may be best to set up a docker network and then use the postgres host name.

This isn't really a libpqxx thing, this is a docker compose networking thing and you'll probably get better answers if you show your docker compose YAML to somebody from the docker compose community together with the commands that you're actually running. If you share them here I can try to help, but I'm no docker compose expert so can probably only help if what you're trying to do is super simple

**Comment by nestorcamposgestal** [comment]
Hi guys, first of all many thanks for the fast responses. I thought it was an issue related with the C++ executable because I can connect to the Database running in the docker from a NodeJS app and from PgAdmin4. 

I share you here the YML file for the docker compose:

```yml
version: "3.9"  # optional since v1.27.0
services:
  database:
    image: postgres
    container_name: stars_gui
    restart: always
    environment: 
      - POSTGRES_PASSWORD=xxx
      - POSTGRES_DB=stars_gui
    ports:
      - "5432:5432"
    volumes:
      - ./docker/database_scripts/ArielDatabase.sql:/docker-entrypoint-initdb.d/ArielDatabase.sql
      - database:/var/lib/postgresql/data
  scheduler:
    build: ./docker
    container_name: ariel_scheduler
    volumes:
      - .:/ariel_scheduler
    working_dir: /ariel_scheduler
    command: mkdir -p build && cd build && cmake ../ && make
    depends_on:
      - database

volumes:
  database:
    driver: local
```

As you can see in the ports command, I do the thing about matching the container's ports to the host machine.

Also, I did try to do a try catch and see some exceptions like this. I don't know if there is a better way for that. But it was giving me the same error.
 
```c++
 try{
        pqxx::connection C("dbname=stars_gui user=postgres password=xxx hostaddr=127.0.0.1 port=5432");

        if(C.is_open()) cout<<"Open"<<endl;
    }
    catch (const std::exception &e)
    {
        std::cerr << e.what() << std::endl;
    }
```

**Comment by jtv** [comment]
Yes, that code looks fine...  I can't think how libpqxx could be involved in the problem, and that makes it hard for me to come up with useful debugging solutions.  But `psql` is a good thing to try, because it's a command-line tool and (like libpqxx) uses libpq to connect to the database.

**Comment by jtv** [comment]
(For the record, the last time we had an issue like this, it did turn out to be something along the lines @KayEss is exploring.)

**Comment by tt4g** [comment]
Do you build and run your C++ applications in a `scheduler` container?
If so, replace `hostaddr=127.0.0.1` with `host=stars_gui`.

You may also want to try changing `pg_hba.conf` to isolate whether the authentication is failing due to different library supported authentication method.
As far as the documentation is concerned, `host all all all scram-sha-256` is set by default in PostgreSQL 14 image and can be changed with the environment variable `POSTGRES_HOST_AUTH_METHOD` (https://hub.docker.com/_/postgres).
e.g. set `POSTGRES_HOST_AUTH_METHOD=md5` then output `host all all all md5`.

**Comment by nestorcamposgestal** [comment]
Hi guys, many thanks for all the responses and the help. Finally I realized what was happening. The thing is that as I was doing a Docker Compose of two containers, one container for the database and another container for my application in C++. And I was running my C++ executable in that docker container where I had all the packages installed for my application. When I was trying connecting to the database I was using the address to localhost (127.0.0.1), but this option only works from my local machine because as you saw in my Docker compose file, I do the map of the ports to from the container to the local machine. But from the container running the C++ app this doesn't work and what I did to make it work is to put in the hostaddr parameter the IP of the docker container with the database.
